### PR TITLE
chore(ci): support pkg preview

### LINKED ISSUE #2082: [Feature Request]: Setup pkg.pr.new
See https://github.com/stackblitz-labs/pkg.pr.new

- [ ] add documentation

### PR DESCRIPTION
**Pull request by shulaoda**
### Description

closes #2082

### CODE REVIEW

**Comment by shulaoda** [comment]
@hyf0 I need your help, please check your email. 😭

**Comment by shulaoda** [comment]
It may be a bit rudimentary and requires further guidance on details.

**Comment by Boshen** [comment]
> _commit: [`84f0d8c`](https://github.com/rolldown/rolldown/runs/29441746198)_
> 
> ```
> pnpm add https://pkg.pr.new/rolldown@2090
> ```
> 
> [Open in Stackblitz](https://pkg.pr.new/template/dd5f8644-2f59-496e-9b01-294fd676168f)

Doesn't work in stackblitz :-(

```
❯ ./node_modules/.bin/rolldown -c
(node:10) ExperimentalWarning: WASI is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
Error: Failed to load native binding
    at file:///home/projects/2nfq8z--run/node_modules/rolldown/dist/shared/plugin-context-data-EkLcm9e5.mjs
    at async _0x247b36._evaluate (https://2nfq8zrun-vbxr.w-credentialless-staticblitz.com/blitz.af4c8d9c.js:40:790218)
    at async ModuleJob.run (https://2nfq8zrun-vbxr.w-credentialless-staticblitz.com/builtins.ddb8d84d.js:155:2451)
    at async ModuleLoader.import (https://2nfq8zrun-vbxr.w-credentialless-staticblitz.com/builtins.ddb8d84d.js:154:2688) {
  [cause]: [
    Error: Cannot find module './rolldown-binding.linux-x64-gnu.node'
    Require stack:
    - /home/projects/2nfq8z--run/node_modules/rolldown/dist/shared/plugin-context-data-EkLcm9e5.mjs
        at Module._resolveFilename (https://2nfq8zrun-vbxr.w-credentialless-staticblitz.com/builtins.ddb8d84d.js:144:12334)
        at Module._load (https://2nfq8zrun-vbxr.w-credentialless-staticblitz.com/builtins.ddb8d84d.js:144:9839)
        at Module.require (https://2nfq8zrun-vbxr.w-credentialless-staticblitz.com/builtins.ddb8d84d.js:144:13097)
        at require3 (https://2nfq8zrun-vbxr.w-credentialless-staticblitz.com/builtins.ddb8d84d.js:163:2815)
        at requireNative (file:///home/projects/2nfq8z--run/node_modules/rolldown/dist/shared/plugin-context-data-EkLcm9e5.mjs:343:14)
        at eval (file:///home/projects/2nfq8z--run/node_modules/rolldown/dist/shared/plugin-context-data-EkLcm9e5.mjs:454:18)
        at eval (file:///home/projects/2nfq8z--run/node_modules/rolldown/dist/shared/plugin-context-data-EkLcm9e5.mjs:52:48)
        at eval (file:///home/projects/2nfq8z--run/node_modules/rolldown/dist/shared/plugin-context-data-EkLcm9e5.mjs:866:32)
        at Generator.next (<anonymous>)
        at _0x20b9f5.run (https://2nfq8zrun-vbxr.w-credentialless-staticblitz.com/blitz.af4c8d9c.js:40:782597) {
      code: 'MODULE_NOT_FOUND',
      requireStack: [Array]
    },
    Error: Cannot load native addon because loading addons is disabled: /home/projects/2nfq8z--run/node_modules/@rolldown/binding-linux-x64-gnu/rolldown-binding.linux-x64-gnu.node
        at _0x1797df._0x1b7d7a [as dlopen] (https://2nfq8zrun-vbxr.w-credentialless-staticblitz.com/blitz.af4c8d9c.js:40:900887)
        at Module._extensions..node (https://2nfq8zrun-vbxr.w-credentialless-staticblitz.com/builtins.ddb8d84d.js:144:15217)
        at Module.load (https://2nfq8zrun-vbxr.w-credentialless-staticblitz.com/builtins.ddb8d84d.js:144:12820)
        at Module._load (https://2nfq8zrun-vbxr.w-credentialless-staticblitz.com/builtins.ddb8d84d.js:144:10273)
        at Module.require (https://2nfq8zrun-vbxr.w-credentialless-staticblitz.com/builtins.ddb8d84d.js:144:13097)
        at require3 (https://2nfq8zrun-vbxr.w-credentialless-staticblitz.com/builtins.ddb8d84d.js:163:2815)
        at requireNative (file:///home/projects/2nfq8z--run/node_modules/rolldown/dist/shared/plugin-context-data-EkLcm9e5.mjs:348:14)
        at eval (file:///home/projects/2nfq8z--run/node_modules/rolldown/dist/shared/plugin-context-data-EkLcm9e5.mjs:454:18)
        at eval (file:///home/projects/2nfq8z--run/node_modules/rolldown/dist/shared/plugin-context-data-EkLcm9e5.mjs:52:48)
        at eval (file:///home/projects/2nfq8z--run/node_modules/rolldown/dist/shared/plugin-context-data-EkLcm9e5.mjs:866:32) {
      code: 'ERR_DLOPEN_DISABLED'
    }
  ]
}

Node.js v18.20.3
```

**Comment by Boshen** [comment]
Works in my local setup (https://github.com/oxc-project/monitor-oxc)

**Review by Boshen — APPROVED** [review]
Let's wait for the core maintainers to take a look at this.

File: .github/workflows/preview-commit.yml [hunk 1]
```diff
@@ -0,0 +1,50 @@
+name: Preview Commit
```
Comment (Boshen) [inline .github/workflows/preview-commit.yml:2]: Need a description linking to https://github.com/stackblitz-labs/pkg.pr.new

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -156,3 +156,16 @@ jobs:
 
       - name: Lint Toml
         run: pnpm lint-toml
+
+  pkg-preview:
```
Comment (Boshen) [inline .github/workflows/ci.yml:160]: Should this run on every pr or just the main branch?

**Comment by shulaoda** [comment]
> Doesn't work in stackblitz :-(

It seems that these dependencies also need to be built, as they were not packaged into the `rolldown` package.

We need to find a way to bind it to the specified source.

```json
  "optionalDependencies": {
    "@rolldown/binding-darwin-x64": "0.12.2",
    "@rolldown/binding-freebsd-x64": "0.12.2",
    "@rolldown/binding-linux-arm-gnueabihf": "0.12.2",
    "@rolldown/binding-darwin-arm64": "0.12.2",
    "@rolldown/binding-linux-arm64-gnu": "0.12.2",
    "@rolldown/binding-linux-arm64-musl": "0.12.2",
    "@rolldown/binding-linux-x64-gnu": "0.12.2",
    "@rolldown/binding-linux-x64-musl": "0.12.2",
    "@rolldown/binding-wasm32-wasi": "0.12.2",
    "@rolldown/binding-win32-arm64-msvc": "0.12.2",
    "@rolldown/binding-win32-ia32-msvc": "0.12.2",
    "@rolldown/binding-win32-x64-msvc": "0.12.2"
  },
```

**Comment by Boshen** [comment]
> We need to find a way to bind it to the specified source.

Building these are going to be really slow. Please make sure we only build on `main` and no where else.

**Comment by shulaoda** [comment]
@Boshen  Package size exceeding 20mb will result in publishing failure.

**Comment by Boshen** [comment]
Intresting, one thing we can do is only build for mac as most of us is on a mac ...

**Comment by shulaoda** [comment]
> Intresting, one thing we can do is only build for mac as most of us is on a mac ...

I will try to apply for a whitelist, but if the application is not approved, I may need to ignore some platforms.

File: packages/rolldown/package.json [hunk 3]
```diff
@@ -126,7 +126,7 @@
     "why-is-node-running": "^3.0.0",
     "zod-to-json-schema": "^3.23.2"
   },
-  "optionalDependencies": {
+  "peerDependencies": {
```
Comment (IWANABETHATGUY) [inline packages/rolldown/package.json]: Why did you replace `optionalDependencies` to `peerDependencies`?

**Comment by Boshen** [comment]
I saw Biome on their users list, can you also take a look at how Biome does this?

File: packages/rolldown/package.json (hunk 3 above)
Comment (shulaoda) [reply packages/rolldown/package.json]: This is a functional test, and I want it to support version replacement for `workspaces`. I have checked its code and it seems that there is some processing for `peerDependencies`, but it doesn't seem to work.

**Comment by shulaoda** [comment]
> I saw Biome on their users list, can you also take a look at how Biome does this?

I am currently exploring and it seems that version replacement for `pnpm workspace` is not supported. I will try to resolve this issue.

**Comment by shulaoda** [comment]
> I saw Biome on their users list, can you also take a look at how Biome does this?

let me take a look.

**Comment by shulaoda** [comment]
> I saw Biome on their users list, can you also take a look at how Biome does this?

They packaged the `wasm` file into one package instead of splitting it apart like we did. I need to ask the repo manager.

**Comment by shulaoda** [comment]
> I saw Biome on their users list, can you also take a look at how Biome does this?

They packaged the `wasm` file into one package instead of splitting it apart like we did. I need to ask the repo manager.

**Comment by shulaoda** [comment]
If not supported, I will try to package all binding files into package `rolldown`.

**Comment by shulaoda** [comment]
Works well! 👀
image

**Review by IWANABETHATGUY — APPROVED** [review]

**Pull request merged**